engine/grid.py

```diff
@@ -200,96 +210,93 @@ def get_neighbor_node(self, row, col, row_max, col_max):
         A1 = area(x, y, x2, y2, x3, y3)
-
+        
         # Calculate area of triangle PAC
         A2 = area(x1, y1, x, y, x3, y3)
-
+        
         # Calculate area of triangle PAB
         A3 = area(x1, y1, x2, y2, x, y)
-
+        
         # Check if sum of A1, A2 and A3
         # is same as A
-        if (A == A1 + A2 + A3):
+        if(A == A1 + A2 + A3):
             return True
         else:
             return False
-
-    # Activate triangle based on three points
-
-    def activate_traingle(self, x1, y1, x2, y2, x3, y3):
-
+    
+    ##Activate a single vertical line of length n with bottom starting at x, y 
+    def activate_line(self,x,y,n):
+        for i in range(n):
+            self.nodes[x,y+i].is_active = True
+
+    ##Activate triangle based on three points
+    def activate_triangle(self,x1,y1,x2,y2,x3,y3):
         """
         Activates all the nodes in a traingle.
         """
-        rows = self.nodes.shape[0]
-        cols = self.nodes.shape[0]
-        for x in range(rows):
-            for y in range(cols):
-                if self.isInsideTriangle(x1, y1, x2, y2, x3, y3, x, y):
-                    self.nodes[x, y].is_active = True
-
-    # Checks if node is on border of activated nodes
-    def is_on_border(self, row, col, row_limit, col_limit):
-        min_col = max(0, col-1)
-        min_row = max(0, row-1)
-        max_col = min(col_limit, col+1)
-        max_row = min(row_limit, row+1)
-
-        # If this node is on the very edge of the grid, it is automatically a border node
-        if min_col == 0 or min_row == 0 or max_col == col_limit or max_row == row_limit:
-            return True
-
-        # If the node has a neighboring node that is inactive, it is a border node
-        for col in range(min_col, max_col+1):
-            for row in range(min_row, max_row+1):
-                if not self.nodes[row][col].is_active:
-                    return True
-        return False
+        for x in range(self.num_rows):
+            for y in range(self.num_cols):
+                if self.isInsideTriangle(x1,y1,x2,y2,x3,y3,x,y):
+                    self.nodes[x,y].is_active = True
+
+    ##Checks if node is on border of activated nodes
+    def is_on_border(self,row, col, row_limit, col_limit):
+            min_col = max(0, col-1)
+            min_row = max(0, row-1)
+            max_col = min(col_limit, col+1)
+            max_row = min(row_limit, row+1)
+
+            # If this node is on the very edge of the grid, it is automatically a border node
+            if min_col == 0 or min_row == 0 or max_col == col_limit or max_row == row_limit:
+                return True
+
+            # If the node has a neighboring node that is inactive, it is a border node
+            for col in range(min_col, max_col+1):
+                for row in range(min_row, max_row+1):
+                    if not self.nodes[row][col].is_active:
+                        return True
+            return False
 
```

Comment:
isn't this only checking if one node is on the border of activated nodes? why is there a loop?